File: Practice/37/Python/example37.py
import math
from enum import IntEnum, auto
class coord_systems(IntEnum):
	Cartesian = auto() # 1
	Polar = auto() # 2

def equal(a, b, e=1E-10):
	if -e < a - b < e: return True
	else: return False

class Point:
	def __init__(self,a1=0,a2=0,coord_system=coord_systems.Cartesian):
		self.deviation = pow(10,-10)
		if type(a1) == str:
			xy = a1.replace('(','').replace(')','').split(',')
			self.x = float(xy[0])
			self.y = float(xy[1])
			self.r = math.hypot(float(xy[0]),float(xy[1]))
			self.phi = math.atan2(float(xy[1]),float(xy[0]))
		else:
			if coord_system == coord_systems.Cartesian: # Если Декартова
				self.x = a1 
				self.y = a2
				self.r = math.hypot(self.x,self.y)
				self.phi = math.atan2(self.y,self.x)
			else: # Если Полярная
				self.r = a1
				self.phi = a2
				self.x = self.r * math.cos(self.phi)
				self.y = self.r * math.sin(self.phi)

	def CartesianToPolar(self):
		self.r = math.hypot(self.x,self.y)
		self.phi = math.atan2(self.y,self.x)

	def PolarToCartesian(self):
		self.x = self.r * math.cos(self.phi)
		self.y = self.r * math.sin(self.phi)

	def __eq__(self, rhs):
		if abs(self.x - rhs.x)<self.deviation and abs(self.y - rhs.y)<self.deviation:
			return 1
		else:
			return 0
	def __ne__(self, rhs): 
		if abs(self.x - rhs.x)<self.deviation and abs(self.y - rhs.y)<self.deviation:
			return 0
		else:
			return 1

	def __str__(self):
		return "(" + str(self.x) + "," + str(self.y) + ")"
	def __repr__(self):
		return "(" + str(self.x) + "," + str(self.y) + ")"
	def get_x(self):
		return self.x
	def get_y(self):
		return self.y

	def get_r(self):
		return self.r
	def get_phi(self):
		return self.phi


	def set_x(self,x):
		self.x = x
		self.CartesianToPolar()

	def set_y(self,y):
		self.y = y
		self.CartesianToPolar()


	def set_r(self,r):
		self.r = r
		self.PolarToCartesian()
	def set_phi(self,phi):
		self.phi = phi
		self.PolarToCartesian()

class Vector:
	def __init__(self,begin=None,end=None):
		if type(begin) != Point and type(end) != Point:
			self.begin = Point(0,0)
			self.end = Point(1,0)
		elif type(begin) == Point and type(end) != Point:
			self.begin = Point(0.0)
			self.end = begin
		else:
			self.begin = begin
			self.end = end
		self.deviation = pow(10,-10)
		self.phi = math.atan2(self.end.get_y() - self.begin.get_y(),self.end.get_x() - self.begin.get_x())
	def __eq__(self, rhs):
		return (self.end.get_x() - self.begin.get_x()) - (rhs.end.get_x() - rhs.begin.get_x()) < self.deviation and (self.end.get_y() - self.begin.get_y()) - (rhs.end.get_y() - rhs.begin.get_y()) < self.deviation
	
	def __neg__(self): 
		return Vector(self.end,self.begin)


	def __add__(self, rhs): 
		return Vector(self.begin,Point(self.end.get_x() + rhs.end.get_x() - rhs.begin.get_x(),self.end.get_y() + rhs.end.get_y() - rhs.begin.get_y()))

	def __sub__(self, rhs):
		return Vector(self.begin,Point(self.end.get_x() - rhs.end.get_x() - rhs.begin.get_x(), self.end.get_y() - rhs.end.get_y() - rhs.begin.get_y()))
	def length(self):
		return math.hypot(self.end.x - self.begin.x, self.end.y - self.begin.y)

	def __mul__(self, rhs):
		if type(rhs) == Vector and type(self) == Vector:
			return	self.length() * rhs.length() * math.cos(self.phi-rhs.phi)
		else:
			return Vector(self.begin,Point(self.begin.get_x() + (self.end.get_x() - self.begin.get_x())*rhs, self.begin.get_y() + (self.end.get_y() - self.begin.get_y()) * rhs))

a = Vector(Point(1, 2))
b = Vector(Point(-2, 0), Point(-1, 2))
if a == b and b == a: print('Equality test passed')
else: print('Equality test failed')
na  = Vector(Point(-1, -2))
ox  = Vector(Point( 1,  0))
nox = Vector(Point(-1,  0))
oy  = Vector(Point( 0,  1))
noy = Vector(Point( 0, -1))
if a == -na and na == -a and -ox == nox and -oy == noy: print('Invert test passed')
else: print('Invert test failed')

if ox + oy + oy == a and -ox == -a + oy + oy: print('Summation test passed')
else: print('Summation test failed')
  
if -ox + oy == oy - ox and -oy + ox == ox - oy: print('Subtraction test passed')
else: print('Subtraction test failed')
  
if (ox * 3 == ox + ox + ox and
    oy * 3 == oy + oy + oy and
    ox * (-3) == -ox - ox - ox and
    oy * (-3) == -oy - oy - oy): print('Multiplication by number test passed')
else: print('Multiplication by number test failed')
  
if (equal(ox.length(), 1) and
    equal(oy.length(), 1) and
    equal((ox * 3 + oy * 4).length(), 5)): print('Length test passed')
else: print('Length test failed')
  
if (equal(ox*ox, ox.length()**2) and
    equal(oy*oy, oy.length()**2) and
    equal((ox*3 + oy*4)*(ox*3 + oy*4), (ox*3 + oy*4).length()**2)): print('Multiplication by Vector test passed')
else: print('Multiplication by Vector test failed')

File: Practice/37/C++/example37/example37/example37.cpp
#include <iostream>
using namespace std;
double sqr(double a);
bool equal(double a, double b, double e = 1E-10);
enum coord_system {
	Cartesian,
	Polar,
};

class Point {
private:
	double deviation = pow(10, -10); // отклонение при сравнении
	double Da1;// x dek
	double Da2;// y dek
	double Pa1;// r polar
	double Pa2;// phi polar
public:

	void ConverteToP(double Da1, double Da2) {
		Pa1 = hypot(Da1, Da2);// r
		Pa2 = atan2(Da2, Da1);//c
	}
	void ConverteToD(double Pa1, double Pa2) {
		Da1 = Pa1 * cos(Pa2);//
		Da2 = Pa1 * sin(Pa2); //
	}
	Point(double a1 = 0, double a2 = 0, coord_system status = Cartesian)
	{
		if (status == Cartesian) { // декартова
			ConverteToP(a1, a2);
			Da1 = a1;
			Da2 = a2;
		}
		else { // полярная
			ConverteToD(a1, a2);
			Pa1 = a1;
			Pa2 = a2;
		}


	}
	bool operator == (const Point& rhs) {
		if (abs(this->Da1 - rhs.Da1) < deviation and abs(this->Da2 - rhs.Da2) < deviation)
		{
			return 1;
		}
		else {
			return 0;
		}
	}
	bool operator != (const Point& rhs) {
		if (abs(this->Da1 - rhs.Da1) < deviation and abs(this->Da2 - rhs.Da2) < deviation)
		{
			return 0;
		}
		else {
			return 1;
		}
	}
	double get_x() const { return Da1; }
	double get_y() const { return Da2; }

	void set_x(double x) {
		Da1 = x;
		// перерасчет полярной
		ConverteToP(Da1, Da2);

	}
	void set_y(double y) {
		Da2 = y;
		//перерасчет полярной
		ConverteToP(Da1, Da2);
	}

	double get_r() const { return Pa1; }
	double get_phi() const { return Pa2; }

	void set_r(double r) {
		Pa1 = r;
		ConverteToD(Pa1, Pa2);
	}
	void set_phi(double phi) {
		Pa2 = phi;
		ConverteToD(Pa1, Pa2);
	}


};
class Vector {

private:
	double deviation = pow(10, -10); // отклонение при сравнении
	Point begin;
	Point end;
public:
	Vector() {
		begin = Point(0, 0);
		end = Point(1, 0);
	}
	Vector(Point end) : end(end) {
		begin = Point(0, 0);
	}
	Vector(Point begin, Point end) : begin(begin), end(end) {
	}
	Point get_begin() const { return begin; }
	Point get_end() const { return end; }
	bool operator == (const Vector& rhs) {
		if ((this->end.get_x() - this->begin.get_x()) - (rhs.end.get_x() - rhs.begin.get_x()) < deviation and (this->end.get_y() - this->begin.get_y()) - (rhs.end.get_y() - rhs.begin.get_y()) < deviation)
		{
			return 1;
		}
		else {
			return 0;
		}
	}
	Vector operator- () {
		Vector p;
		p.begin = this->end;
		p.end = this->begin;
		return p;
	}
	Vector operator+ (const Vector& rhs) {
		Vector p;
		p.begin = this->begin;
		p.end = Point(this->end.get_x() + rhs.end.get_x() - rhs.begin.get_x(), this->end.get_y() + rhs.end.get_y() - rhs.begin.get_y());
		return p;
	}
	Vector operator- (const Vector& rhs) {
		Vector p;
		p.begin = this->begin;
		p.end = Point(this->end.get_x() - rhs.end.get_x() - rhs.begin.get_x(), this->end.get_y() - rhs.end.get_y() - rhs.begin.get_y());
		return p;
	}
	Vector operator* (double rhs) {
		Vector p;
		p.begin = this->begin;
		p.end = Point(this->begin.get_x() + (this->end.get_x() - this->begin.get_x()) * rhs, this->begin.get_y() + (this->end.get_y() - this->begin.get_y()) * rhs);
		return p;
	}
	double length() {
		return hypot(this->end.get_x() - this->begin.get_x(), this->end.get_y() - this->begin.get_y());
	}
	double operator* (const Vector& rhs) {
		double rhslen = hypot(rhs.end.get_x() - rhs.begin.get_x(), rhs.end.get_y() - rhs.begin.get_y());
		double scalar = (this->length()) * rhslen * cos(atan2((this->end.get_y() - this->begin.get_y()), (this->end.get_x() - this->begin.get_x())) - atan2((rhs.end.get_y() - rhs.begin.get_y()), (rhs.end.get_x() - rhs.begin.get_x())));
		return scalar;
	}
};



int main()
{
	Vector a(Point(1, 2)), b(Point(-2, 0), Point(-1, 2));


	if (a == b && b == a) cout << "Equality test passed\n";
	else cout << "Equality test failed\n";

	Vector na(Point(-1, -2)), ox(Point(1, 0)), nox(Point(-1, 0)), oy(Point(0, 1)), noy(Point(0, -1));
	if (a == -na && na == -a && -ox == nox && -oy == noy) cout << "Invert test passed\n";
	else cout << "Invert test failed\n";

	if (ox + oy + oy == a && -ox == -a + oy + oy) cout << "Summation test passed\n";
	else cout << "Summation test failed\n";

	if (-ox + oy == oy - ox && -oy + ox == ox - oy) cout << "Subtraction test passed\n";
	else cout << "Subtraction test failed\n";

	if (ox * 3 == ox + ox + ox &&
		oy * 3 == oy + oy + oy &&
		ox * (-3) == -ox - ox - ox &&
		oy * (-3) == -oy - oy - oy) cout << "Multiplication by number test passed\n";
	else cout << "Multiplication by number test failed\n";

	if (equal(ox.length(), 1) &&
		equal(oy.length(), 1) &&
		equal((ox * 3 + oy * 4).length(), 5)) cout << "Length test passed\n";
	else cout << "Length test failed\n";

	if (equal(ox * ox, sqr(ox.length())) &&
		equal(oy * oy, sqr(oy.length())) &&
		equal((ox * 3 + oy * 4) * (ox * 3 + oy * 4), sqr((ox * 3 + oy * 4).length()))) cout << "Multiplication by Vector test passed\n";
	else cout << "Multiplication by Vector test failed\n";
}

bool equal(double a, double b, double e) {
	if (-e < a - b && a - b < e) return true;
	else return false;
}

double sqr(double a) {
	return a * a;
}